### ISSUE TITLE
Python 3 with Python2?

### ISSUE THREAD

**Issue by DavidMcKenna** [bitwiseworks/python-os2#6]
Decided to give Python3 a try (on the -exp repository) and when I try to install (with ANPM 1.06) I get this odd message:

python2-libcomps-0.1.15-1.oc00.pentium4 requires python(abi) = 2.7

  I have python2-libcomps already installed - is it saying I can't use Python3? Can Python3 be installed along with Python 2.7.6?

 Also, I tried to install Python2.7.18 (no update for 2.7.6?) but I get this also odd message:

 No available packages for the following names were found.
   python2

 Not sure how I can manage to uninstall 2.7.6 to then install 2.7.18 (at least if using YUM).

**Comment by SilvanScherrer** [comment]
In theory (and on all my systems) python3 is an update for python and python2. If you do yum check-update you should see it. And I don't see the python2-libcomps issue. Try to deinstall python2-libcomps and then try to update again.

**Comment by DavidMcKenna** [comment]
If I uninstall python2-libcomps, then try to install python3, I now get this error:

yum-utils-1.1.31-3.oc00.noarch requires python(abi) = 2.7

 So it seems there are other items tied to python 2.7

 Yum check-update shows:

[C:\]yum check-update
Loaded plugins: fastestmirror, replace
Loading mirror speeds from cached hostfile
Obsoleting Packages
python-unversioned-command.pentium4          3.9.5-3.oc00           netlabs-exp

    python.pentium4                          2.7.6-25.oc00          installed

python3-libs.pentium4                        3.9.5-3.oc00           netlabs-exp

    python-libs.pentium4                     2.7.6-25.oc00          @netlabs-rel

 Not clear to me what this means, but python3 does not show as an 'update'....

**Comment by SilvanScherrer** [comment]
David
all is good. In theory (which seems to not the case on your system) python3 is an update and also installs python2. So both are there in parallel. Why this is not working on your systems is to be found out. Right now I have no idea, as on all my test systems and some other testers it worked :( I let you know if I need more info.
Ah one thing could be good. "yum list installed"

**Comment by DavidMcKenna** [comment]
OK, thanks for the explanation. Here is the result of 'yum list installed'.
[install.txt](https://github.com/bitwiseworks/python-os2/files/6623237/install.txt)

**Comment by SilvanScherrer** [comment]
thx for it. please grab also a yum list python* for me. As I still have no clue why you get those issues/messages.

**Comment by DavidMcKenna** [comment]
Here ya go...
[pylist.txt](https://github.com/bitwiseworks/python-os2/files/6627180/pylist.txt)

**Comment by SilvanScherrer** [comment]
if you have a pc to play with. Could you just install arcaos and see if then the update would work? As I still don't know which rpm is quilty.

**Comment by DavidMcKenna** [comment]
I have a couple of laptops I can play around with... I'll try when I can (later in the week) and let you know...

**Comment by DavidMcKenna** [comment]
Tried on my HP laptop, which has all the same updates as my desktop including arcanoae-exp repository, but 'yum update python' didn't work there either (same errors). Went to my server, which does not have any arcanoae-exp updates, but 'yum install arcanoae-exp' doesn't seem to work anymore. Any idea how to add the experimental repository?

**Comment by DavidMcKenna** [comment]
Oops... should be netlabs-exp, and it was installed. But when I 'yum update python' on the server (that has had no experimental files added) I get the same errors as on the desktop reported here...

**Comment by SilvanScherrer** [comment]
could you eventually try on a new AroaOS installation? Where you did not install any additional rpm's? Eventually this gives us a clue which rpm is guilty. As right now i'm on the blind side somehow.

**Comment by DavidMcKenna** [comment]
Yes, I will try that when I can get to it...

**Comment by DavidMcKenna** [comment]
OK, did a fresh install of ArcaOS 5.0.6 on an HP EliteBook 8440p on a 150GB partition. Everything works (except WiFi). After install, I opened a command prompt and typed 'yum install netlabs-exp'.  When that was complete, I typed 'yum update python' and got similar errors as already reported. Output attached...
[newinst.txt](https://github.com/bitwiseworks/python-os2/files/6677822/newinst.txt)

**Comment by SilvanScherrer** [comment]
this issue now really gets me angry ;-) Lets try something. Disable netlabs-exp and update all packages. Then enable netlabs-exp again and try to update python. The strange part is, that all my tester succeeded and on the AOS tester list all seemed to succeed as well. So I must overlook something.

**Comment by DavidMcKenna** [comment]
Sorry, not trying to make you angry, just trying to figure this out. I did as you requested, and it seemed to work - on the Elitebook, after upgrading all items from arcanoae-rel and netlabs-rel (and rebooting), then installing netlabs-exp again, I did not get any error messages when upgrading python, but I did not tell it to install in order to be able to further test adding things that might make it break. Any preference for what I try adding and in what order? One thing I think I might do is re-install the YUM items installed by the ArcaOS installer on my desktop machine in case something is corrupted somehow...

 I have also thought about joining the AOS testers, but my job situation makes me unreliable for that - at times I would be unavailable for long periods, so I just test what I can when I can...

**Comment by SilvanScherrer** [comment]
Not you made me angry dont worry. The Situation that i dont finde the issue does. That it runs after all rel updates is a good progress.

**Comment by DavidMcKenna** [comment]
Spent some time on this today, but still no progress. Here is what I have tried so far...

 I decided to work exclusively with my server machine (with server components disabled) because it is the closest to a fresh install of any of my ArcaOS installations (except of course for the laptop that I just installed). To remind, I can update python on the laptop, but not on the server.

 First thing is, on the server there are these 3 repositories installed: netlabs-rel, arcanoae-rel, and arcanoae-sub. The '-sub' is installed because I need the epson printer drivers available there. Installed the '-sub' repository on the laptop and installed the epson drivers, then rebooted and checked if I can still update python - I can, so no problem there.

 Next I added packages found on the laptop fresh install that were not installed on the server - brlaser and hplip and associated dependencies. Must have deleted those at some point because I do not need them, but we want to have exactly the same packages and versions installed on both laptop and server. Rebooted the server, but still get the errors when trying to update python.
  Also, the server had 'libstdc++-fwd' installed but not on the laptop, but the laptop had 'gcc-stdc++-shared-library' installed but not the server. Found that these both contain a file of the same name and so I can't install both - it has to be one or the other. A lot of things are dependent on these, any idea which is the one I really want to have? Could not fix this.

 Then I removed all packages on the server that were not installed on the laptop. Mostly multimedia stuff not used really. Rebooted the server, tried to update python - no go.

 At this point both the server and laptop have an identical list of installed packages. The only difference is the server has pentium4 versions, and the laptop has i686 versions. And of course when netlabs-exp is installed on both, the laptop will upgrade python, but the server will not.

 Now I went and downloaded all the ArcaNoae driver and app updates and installed them on the freshly installed laptop. After which, I still can upgrade python on the laptop, so no trouble from these. Now both server and laptop have identical OS files installed.

 Finally, I re-installed every YUM package on the server. A few failed to reinstall: cups, and all 3 of the poppler-legacy-xx packages. Cups gives this error:
Error unpacking rpm package 1:cups-2.2.12-1.oc00.pentium4
error: unpacking of archive failed on file /@unixroot/var/run/cups/certs: cpio:
unlink

 and the poppler packages say 'Access denied'. I tried 'yum reinstall cups' on my laptop just to check, and I get the same error there. Same with the poppler packages. Seems those packages have a problem?

 So now the only difference I know of between these 2 computers is the file 'stdcpp.dll' and the pentium4 vs. i686 issue. I copied stdcpp.dll from the laptop to the server, rebooted and still can't update python there. I guess the next thing is to update all the files on the laptop to pentium4 versions to see if/when 'yum upgrade python' breaks...

**Comment by DavidMcKenna** [comment]
One thing I didn't mention is the server was originally installed as ArcaOS 5.0.4, and then updated to 5.0.5, then again to 5.0.6. Doesn't seem like it should matter, but just putting that out there...

**Comment by DavidMcKenna** [comment]
OK, been slowly updating the laptop's YUM packages to pentium4 versions and found an issue immediately: if I set the platform to Pentium4 and have netlabs-exp installed, then the python update fails as previously described, but if set to i686 it succeeds. So to get a successful python update, the platform must be i686. But that does not help on the server, it still fails even with the platform set to i686 (although all it's packages are pentium4 versions).

 On the laptop I have been updating to Pentium4 in alphabetical order (using ANPM), and some packages do not update even though they show a Pentium4 version (so far: coreutils, cups-filters, cups-filters-libs, dash, dash-sh, fontconfig, and freetype), so those have not been changed. But I found that changing Python 2.7.6-25 to the Pentium4 version (and it's dependencies) will also cause the update to fail (even with platform i686). Re-installing the i686 version, the update will succeed. But again on the server, setting Python to the i686 version, and platform i686 with netlabs-exp still does not allow python to be updated. So there is still something else going on....

**Comment by SilvanScherrer** [comment]
we appreciate all your effort. we need to read all what you wrote. eventually we find some issues. what we know, is, that changing the arch is never easy. So this might lead to strange messages.

**Comment by DavidMcKenna** [comment]
Some progress...

 Continued replacing the i686 versions of packages on the laptop with pentium4 ones and got through the whole list. The packages 'python, rpm, and yum' must be i686 versions or the 'yum update python' will not work (along with i686 as the platform). Several packages would not allow changing to the pentium4 version though- here is the entire list:

coreutils, cups-filters, cups-filters-libs, dash, dash-sh, fontconfig, freetype, lcms2, libvpx-legacy-2, libxml2, mmap, nspr, pthread, sqlite, zlib

 When trying to change, there is no error, it just reinstalls the i686 version. I was able to work around the issue with some of them by downgrading to a previous pentium4 version, then upgrading to the current version. After that the list of packages still at i686 is:

 coreutils, cups-filters, cups-filters-libs, dash-sh, freetype, sqlite

 Went to the server and changed 'python, rpm and yum' packages to i686 versions, but 'yum update python' would still not work as already described. So one of the above listed i686 packages must need to be changed.

 It turned out I can change all of them to i686 packages with no problem, but not change them back to pentium4. No matter, when I changed sqlite to the i686 package on the server, 'yum update python' started to work! So the full list of i686 packages I need in order to update python on the server are 'python, rpm, yum, and sqlite'.

 Thinking I had it solved, I went to my desktop and changed those packages and platform to i686, but yum update python wouldn't work there. Of course I have dozens of experimental packages installed on that machine, so need to figure out which ones are interfering <sigh>. This machine is where I normally do all my testing, so would prefer to install the new python here....

**Comment by SilvanScherrer** [comment]
to be honest I still have no idea why it's not working for you. I did a couple of installs/updates and they all worked. I'm a bit lost here.

**Comment by DavidMcKenna** [comment]
No ideas here either.. I have tried to find a work-around, but so far no luck on my main desktop machine.

**Comment by DavidMcKenna** [comment]
You were able to update from a fresh install after changing the platform to Pentium4? I have not yet been able to do that here... it must be set to i686 to work. Once I have the new pythons installed that way, I can set platform to Pentium4 and update them to Pentium4 too.

**Comment by DavidMcKenna** [comment]
I finally got this to work - here's how:

 Decided to disable all repositories on my desktop EXCEPT netlabs-exp, and tried 'yum update python'. When I did this, it ended with an error: 'can't find pkg-config.exe', and indeed, I did not have a 'pkg-config.exe' on my desktop system (but I looked at one of the laptops where I could update python and it was there). So I re-enabled the netlabs-rel repository and installed the 'pkgconfig' package, then disabled 'netlabs-rel' and retried 'yum update python'. It worked! Even set to Pentium4 platform!

 So I guess the 'pkgconfig' package is a requirement for the new python, and maybe there is some conflict with the 2 netlabs repositories? In any case I am running the new pythons - thanks for making them available.

**Comment by DavidMcKenna** [comment]
I should mention that if both netlabs-rel and netlabs-exp were enabled, then I could not update...

**Comment by SilvanScherrer** [comment]
this is good on one hand, but very bad on the other. As I still don't get why it's not working when both repos are enabled. As this has to work in theory.

Yes pkgconfig is a req for python and many other packages.

You are welcome

**Comment by DavidMcKenna** [comment]
One thing I just realized is that on the desktop machine I had the most trouble with, I always add the -debuginfo package of everything I install (and never do that on any other system). I just went to install the -debuginfo packages of python2.7 and python3 and got an error saying it conflicted with python-debuginfo. Uninstalled python-debuginfo and installed python2.7-debuginfo, then python3-debuginfo would not install (saying it conflicted with python2.7-debuginfo). So both the python2.7 and python3 -debuginfo packages cannot co-exist. Could the presence of the -debuginfo packages be related to the install issues I saw?
